crates/driver/src/domain/competition/mod.rs

```diff
@@ -328,6 +302,127 @@ impl Competition {
+        .await
+        .is_err()
+    {
+        tracing::warn!("reached timeout while encoding");
+    }
+
+    let mut settlements = vec![];
+    while let Ok(settlement) = receiver.try_recv() {
+        settlements.push(settlement);
+    }
+    settlements
+}
+
+/// Tries to merge multiple settlements into a single settlement.
+/// If merged, the merged settlement is added to the `settlements` vector and
+/// the contained settlements are removed from `settlements` vector.
+/// If the deadline is reached, current state of `settlements` vector is
+/// returned.
+async fn merge_settlements(
+    settlements: &mut Vec<Settlement>,
+    eth: &Ethereum,
+    simulator: &Simulator,
+    deadline: auction::Deadline,
+) {
+    if settlements.len() <= 1 {
+        // Nothing to merge.
+        return;
+    }
+
+    // TODO(#1483): parallelize this
+    // TODO(#1480): more optimal approach for settlement merging
+
+    // Merge the settlements in random order.
+    settlements.shuffle(&mut rand::thread_rng());
+
+    // The merging algorithm works as follows: the [`inner_settlements`] vector
+    // keeps the "most merged" settlements until they can't be merged anymore,
+    // at which point they are moved into the [`settlements`] vector.
+
+    let task = async {
+        // work on the copy to make sure the timeout doesn't leave original in an
+        // inconsistent state
+        let mut inner_settlements = settlements.clone();
+
+        while let Some(settlement) = inner_settlements.pop() {
+            // Has [`settlement`] been merged into another settlement?
+            let mut merged = false;
+
+            // Try to merge [`settlement`] into some other settlement.
+            for other in inner_settlements.iter_mut() {
+                match other.merge(&settlement, eth, simulator).await {
+                    Ok(m) => {
+                        *other = m;
+                        merged = true;
+                        observe::merged(&settlement, other);
+                        break;
+                    }
+                    Err(err) => {
+                        observe::not_merged(&settlement, other, err);
+                    }
+                }
+            }
+
+            // If [`settlement`] can't be merged into any other settlement, this is its
+            // final, most optimal form.
+            if !merged {
+                // remove all settlements from the `settlements` vector that are substituted
+                // with the [`settlement`]
+                settlements.retain(|other| {
+                    other
+                        .solutions()
+                        .iter()
+                        // it's enough to check for one solution id, since either all ids are contained in [`settlement`] or none
+                        .next()
+                        .map(|s| !settlement.solutions().contains(s))
+                        .unwrap_or(true)
+                });
```

Comment:
I personally find this logic quite complex (significantly more so than the existing one). Is there any harm in not removing the original solutions given that we allow to return a list?